### PR TITLE
rustdoc-search: show type signature on type-driven SERP

### PR DESCRIPTION
**Pull request by notriddle**
This is a rewrite of https://github.com/rust-lang/rust/pull/117112, based on feedback by @aDotInTheVoid.

Part of https://github.com/rust-lang/rust-project-goals/issues/112

# Preview

Preview page:

* https://notriddle.com/rustdoc-html-demo-11/type-signature-v2/std/index.html?search=vec%3CU%3E%20-%3E%20%5BT%5D
* https://notriddle.com/rustdoc-html-demo-11/type-signature-v2/std/index.html?search=option%3Ct%3E%20-%3E%20result%3Ct%3E
* https://notriddle.com/rustdoc-html-demo-11/type-signature-v2/std/index.html?search=Vec%3CAllocator%3E%20-%3E%20Box%3C%5BT%5D%3E
* https://notriddle.com/rustdoc-html-demo-11/type-signature-v2/std/index.html?search=Iterator%3CT%3E%2C%20(T->bool)%20-%3E%20(Extend%3CT%3E%2C%20Extend%3CT%3E)
* https://notriddle.com/rustdoc-html-demo-11/type-signature-v2/std/index.html?search=new%20-%3E%20HashMap%3CK%2C%20V%2C%20S%3E (this is an old "bug report" from @Urgau — the results aren't any different, but the visible type signature and headline message makes it more obvious what's going on)

Screenshots:

image

image

image

image

image

image

# Rationale

Since the point of this feature is to answer the question "why this a match," it's important that it communicates a few things:

* **How do I use this?** This UI shows types as Rustdoc sees them, using Rustdoc's search syntax, rather than the syntax used by normal Rust. It's supposed to act as a built-in, context-sensitive tutorial, since you can type a name in, switch to the In Parameters tab, and see what you could've typed to get a particular result.
* **Which type parameters got matched to other type parameters?** This information is found in a popover that you can open to see information about a search result. This makes the visible type signature a strict dump of the search index, providing reasonably clear separation of user input from output.
* **How does the search engine even work, anyway?** It's important that this page accurately shows you what unboxing occurred and which mappings got followed. That's why parts of the highlighter are baked into the type unification function: it doesn't just show you where your atoms occur, but actually shows you which ones *got matched*. A highlighted "where clause" is also found in the About This Result popover.

This change doesn't touch deduplication. That logic needs some adjusting, but is separate enough from this problem to be done in a different PR.

# How it works

A particular source of complexity in this PR is caused by storing the names of type parameters. Since this data is only needed after the results come up, loading them gets delayed until after the search is done, so that they can be loaded at the same time as the descriptions.

# Future possibilities

This PR is based on https://github.com/rust-lang/rust/pull/124148. This feature benefits from the type signature index itself being accurate, so we really kinda need full support for Rust's type grammar.

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
I'll let @aDotInTheVoid confirms it looks good before reviewing the code.

**Comment by aDotInTheVoid** [comment]
Having played around with this, I really like it. I think the searching side seems really solid and have some small nitpicks on the UI side. Feel free to disregard them, as I'm not great at this, and others on the rustdoc team have put alot more thaught into our UI than I have.

In rough order of importance:

- The type signature should probably be in monospace. Seeing "Vec\<u8\> -> String" is much weirder to me than "`Vec<u8> -> String`"
- I don't think the "About this result" infobox worth having in most (if not all cases), now that we use the source name for generics. It should defiantly be elided when it only had info like [ `T is T` ](https://notriddle.com/rustdoc-html-demo-11/type-signature-v2/std/index.html?search=T%20-%3E%20Box%3CT%3E)
  - Additionally, I'm not sure how nessesary it it to display the bounds here. Is saying [`K: HashEq`](https://notriddle.com/rustdoc-html-demo-11/type-signature-v2/std/string/struct.String.html?search=%5B(K%2C%20V)%5D%20-%3E%20HashMap%3CK%2C%20V%3E#method.from_utf8_unchecked) helpfull here
    - Random aside: In the infobox they say `K: HashEq`, but have [`K: Hash + Eq`](https://notriddle.com/rustdoc-html-demo-11/type-signature-v2/src/std/collections/hash/map.rs.html#1360-1376)  
- Really Small Thing: the (i) for [notable traits](https://notriddle.com/rustdoc-html-demo-11/type-signature-v2/std/iter/trait.Iterator.html#method.flatten) changes color on highlight. 
- I don't love  [`-> Vec<T>`](https://notriddle.com/rustdoc-html-demo-11/type-signature-v2/std/collections/hash_map/struct.HashMap.html?search=-%3E%20Vec%3CT%3E#method.from) for a function that takes no arguments, but I don't have better ideas.


Thanks again for working on this, I'm really excited about having this!

File: src/librustdoc/html/static/css/rustdoc.css [hunk 1]
```diff
@@ -2743,17 +2748,19 @@ Original by Dempfi (https://github.com/dempfi/ayu)
 	border-right: 1px solid #ffb44c;
 }
 
-:root[data-theme="ayu"] .search-results a:hover,
-:root[data-theme="ayu"] .search-results a:focus {
+:root[data-theme="ayu"] .search-results li:hover > a,
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/css/rustdoc.css:2751]: Having this special case for the ayu theme is really not great, we should try to either make a new variable or "fix" the ayu theme at some point (not for this PR).

File: src/librustdoc/html/static/js/main.js [hunk 2]
```diff
@@ -353,6 +353,27 @@ function preLoadCss(cssUrl) {
         loadedDescShard: function(crate, shard, data) {
             this.descShards.get(crate)[shard].resolve(data.split("\n"));
         },
+        paramNameShards: new Map(),
+        paramNameResolvers: new Map(),
+        loadParamNames: async function(crate) {
+            if (this.paramNameShards.has(crate)) {
+                return this.paramNameShards.get(crate);
+            } else {
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/main.js:361]: You can remove this `else` since you already returned. It'll allow to remove one indent level.

File: src/librustdoc/html/static/js/search.js [hunk 3]
```diff
@@ -1382,16 +1411,291 @@ function initSearch(rawSearchIndex) {
...
+            /** @param {QueryElement} queryElem */
+            function remapQuery(queryElem) {
+                if (queryElem.id < 0) {
+                    queryParamNames[-1 - queryElem.id] = queryElem.name;
+                }
+                if (queryElem.generics.length > 0) {
+                    queryElem.generics.forEach(remapQuery);
+                }
+                if (queryElem.bindings.size > 0) {
+                    [...queryElem.bindings.values()].flat().forEach(remapQuery);
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:1501]: Can't you do `queryElem.bindings.values().forEach(remapQuery);`?

File: src/librustdoc/html/static/js/search.js [hunk 4]
```diff
@@ -1382,16 +1411,291 @@ function initSearch(rawSearchIndex) {
...
+                }
+            }
+            parsedQuery.elems.forEach(remapQuery);
+            parsedQuery.returned.forEach(remapQuery);
+            function pushText(fnType, result) {
+                // If !!(result.length % 2) == false, then pushing a new slot starts an even
+                // numbered slot. Even numbered slots are not highlighted.
+                if (!!(result.length % 2) === !!fnType.highlighted) {
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:1509]: This code is very complicated to understand. Why do you need `!!`?

File: src/librustdoc/html/static/js/search.js [hunk 5]
```diff
@@ -1382,16 +1411,291 @@ function initSearch(rawSearchIndex) {
...
+                    result.push("");
+                } else if (result.length === 0 && !!fnType.highlighted) {
+                    result.push("");
+                    result.push("");
+                }
+                result[result.length - 1] += fnType.name;
+            }
+            function writeHof(fnType, result) {
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:1517]: Could you add a code comment explaining what's going on in this function please?

File: src/librustdoc/html/static/js/search.js [hunk 6]
```diff
@@ -1382,16 +1411,291 @@ function initSearch(rawSearchIndex) {
...
+                }
+                if (hofOutput.length > 1) {
+                    pushText({name: ")", highlighted: false}, result);
+                }
+            }
+            /**
+             * @param {FunctionType} fnType
+             * @param {[string]} result
+             */
+            function writeFn(fnType, result) {
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:1553]: A comment explaining what this function does would be welcome too.

File: src/librustdoc/html/static/js/search.js [hunk 7]
```diff
@@ -2015,32 +2388,18 @@ function initSearch(rawSearchIndex) {
...
-                        elem,
-                        whereClause,
-                        mgens,
-                        unboxingDepth,
-                    );
-                }
+            if (row.id > 0 && elem.id > 0 && elem.pathWithoutLast.length === 0 &&
+                row.generics.length === 0 && elem.generics.length === 0 &&
+                row.bindings.size === 0 && elem.bindings.size === 0 &&
+                // special case
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:2394]: Special case for what?

**Comment by GuillaumeGomez** [comment]
Please also add a test for clicking on the `i` icon in the search results.

One final thought: I think, from what I understood of the code, that maybe it could be simplified to handle the notable trait part. I'll need to come back to this to ensure I didn't miss something obvious.

File: src/librustdoc/html/static/js/search.js (hunk 4 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:1509]: A comment would be a nice addition. :)

File: src/librustdoc/html/static/js/search.js (hunk 7 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:2394]: Please add a comment then. ;)

File: src/librustdoc/html/static/js/main.js [hunk 8]
```diff
@@ -1240,8 +1263,10 @@ function preLoadCss(cssUrl) {
         }
     }
 
-    onEachLazy(document.getElementsByClassName("tooltip"), e => {
-        e.onclick = () => {
+    window.rustdocConfigureTooltip = e => {
+        e.addEventListener("click", ev => {
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/main.js:1267]: We originally made it `onclick` to prevent to add more `click` events iirc. But beyond that, is this change necessary? You still get the event as first argument even with `.onclick = ev => {`.

File: src/librustdoc/html/static/js/search.js (hunk 3 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:1510]: Ah indeed. JS API are weird.

File: src/librustdoc/html/static/js/search.js [hunk 9]
```diff
@@ -1382,16 +1411,330 @@ function initSearch(rawSearchIndex) {
...
+                    const where = [];
+                    let first = true;
+                    for (const nested of fnType.generics) {
+                        if (first) {
+                            first = false;
+                        } else {
+                            pushText({ name: " + ", highlighted: false }, where);
+                        }
+                        writeFn(nested, where);
+                    }
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:1613]: It drives me nuts that we can't skip the first n elements in a `for` loop in JS... Maybe the loop should be rewritten to handle this case (code is shorter too :smiley:):

```suggestion
                    for (let i = 1; len = fnType.generics.length; i < len; ++i) {
                        pushText({ name: " + ", highlighted: false }, where);
                        writeFn(fnType.generics[i], where);
                    }   
```

File: src/librustdoc/html/static/js/search.js [hunk 10]
```diff
@@ -1382,16 +1411,330 @@ function initSearch(rawSearchIndex) {
...
+            }
+            /**
+             * Write a type. This function checks for special types,
+             * like slices, with their own formatting. It also handles
+             * updating the where clause and generic type param map.
+             *
+             * @param {FunctionType} fnType
+             * @param {[string]} result
+             */
+            function writeFn(fnType, result) {
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:1586]: Might be worth splitting this function (especially the `else` part).

File: src/librustdoc/html/static/js/search.js (hunk 9 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:1613]: I just realized that this case was present in a few places. So might be worth writing a function:

```js
function pickAName(array, notFirstElementCall, everyElementsCall) {
    if (array.length !== 0) {
        everyElementsCall(array[0]);
        for (let i = 0, len = array.length; i < len; ++i) {
            const elem = array[i];
            notFirstElementCall(elem);
            everyElementsCall(elem);
        }
    }
}
```

Then we could replace the loop with:

```js
pickAName(fnType.generics, elem => writeFn(elem, where), elem => pushText({ name: " + ", highlighted: false }, where));
```

Well, tell me what you think about it.

File: src/librustdoc/html/static/js/search.js [hunk 11]
```diff
@@ -1382,16 +1411,330 @@ function initSearch(rawSearchIndex) {
...
+                                if (needsSpace) {
+                                    pushText({
+                                        name: " ",
+                                        highlighted: prevHighlighted && value.highlighted,
+                                    }, result);
+                                }
+                                needsSpace = true;
+                                prevHighlighted = value.highlighted;
+                                writeFn(value, result);
+                            }
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:1650]: Similar case as above.

File: src/librustdoc/html/static/js/search.js [hunk 12]
```diff
@@ -1382,16 +1411,330 @@ function initSearch(rawSearchIndex) {
...
+                                ["(", ")"];
+                            pushText({ name: ob, highlighted: fnType.highlighted }, result);
+                            let needsComma = false;
+                            for (const value of fnType.generics) {
+                                if (needsComma) {
+                                    pushText({ name: ", ", highlighted: false }, result);
+                                }
+                                needsComma = true;
+                                writeFn(value, result);
+                            }
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:1633]: Same here.

File: src/librustdoc/html/static/js/search.js [hunk 13]
```diff
@@ -1382,16 +1411,330 @@ function initSearch(rawSearchIndex) {
...
+                                highlighted: false,
+                            }, result);
+                            let needsPlus = false;
+                            for (const value of values) {
+                                if (needsPlus) {
+                                    pushText({ name: " + ", highlighted: false }, result);
+                                }
+                                needsPlus = true;
+                                writeFn(value, result);
+                            }
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:1686]: Here too.

File: src/librustdoc/html/static/js/search.js [hunk 14]
```diff
@@ -1382,16 +1411,330 @@ function initSearch(rawSearchIndex) {
...
+                        }
+                    }
+                    if (fnType.generics.length > 0) {
+                        for (const value of fnType.generics) {
+                            if (needsComma) {
+                                pushText({ name: ", ", highlighted: false }, result);
+                            }
+                            needsComma = true;
+                            writeFn(value, result);
+                        }
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:1699]: Here too.

File: src/librustdoc/html/static/js/search.js [hunk 15]
```diff
@@ -1382,16 +1411,330 @@ function initSearch(rawSearchIndex) {
...
+            }
+            const type = [];
+            let needsComma = false;
+            for (const fnType of fnInputs) {
+                if (needsComma) {
+                    pushText({ name: ", ", highlighted: false }, type);
+                }
+                needsComma = true;
+                writeFn(fnType, type);
+            }
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:1714]: Here too.

File: src/librustdoc/html/static/js/search.js [hunk 16]
```diff
@@ -1382,16 +1411,330 @@ function initSearch(rawSearchIndex) {
...
+            pushText({ name: " -> ", highlighted: false }, type);
+            needsComma = false;
+            for (const fnType of fnOutput) {
+                if (needsComma) {
+                    pushText({ name: ", ", highlighted: false }, type);
+                }
+                needsComma = true;
+                writeFn(fnType, type);
+            }
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:1723]: Here too.

File: src/librustdoc/html/static/js/search.js [hunk 17]
```diff
@@ -2662,31 +3065,32 @@ function initSearch(rawSearchIndex) {
      * @param {Array<?>}    array   - The search results for this tab
      * @param {ParsedQuery} query
      * @param {boolean}     display - True if this is the active tab
+     * @param {"sig"|"elems"|"returned"|null} typeInfo
      */
     async function addTab(array, query, display) {
         const extraClass = display ? " active" : "";
 
-        const output = document.createElement("div");
+        let output = document.createElement("ul");
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:3073]:
```suggestion
        const output = document.createElement(array.length === 0 && query.error === null ? "div" : "ul");
```

File: src/librustdoc/html/static/js/search.js [hunk 18]
```diff
@@ -2702,12 +3106,90 @@ ${item.displayPath}<span class="${type}">${name}</span>\
...
+                };
+                return li;
+            }));
+            lis.then(lis => {
+                for (const li of lis) {
+                    output.appendChild(li);
+                }
+            });
         } else if (query.error === null) {
+            output = document.createElement("div");
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:3192]: Can be removed from change above.

**Comment by GuillaumeGomez** [comment]
So we're now in the "nit picking" review round. This is great work! Can't wait for this to be merged. :)

File: src/librustdoc/html/static/js/search.js (hunk 9 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:1613]: The `bindings` map?

File: src/librustdoc/html/static/js/search.js [hunk 19]
```diff
@@ -1382,16 +1425,324 @@ function initSearch(rawSearchIndex) {
...
+                        writeHof(fnType, result);
+                        return;
+                    }
+                    pushText(fnType, result);
+                    if (fnType.bindings.size > 0 || fnType.generics.length > 0) {
+                        pushText({ name: "<", highlighted: false }, result);
+                    }
+                    if (fnType.bindings.size > 0) {
+                        onEachBtwn(
+                            fnType.bindings,
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:1688]: I think it's fine to exclude `bindings` from `onEachBtwn` so the function can be a bit "better" (not having the `if` condition in the loop)

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:1688]: Fine by me. Btw, did you bench with the JS changes?

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:1688]: There seems to have a noticeable diff:

```
new:

Testing T -> U ... in_args = 0, returned = 0, others = 200
wall time = 1179
Testing T, U ... in_args = 0, returned = 0, others = 200
wall time = 429
Testing T -> T ... in_args = 0, returned = 0, others = 200
wall time = 555
Testing crc32 ... in_args = 0, returned = 0, others = 0
wall time = 655
Testing spi::pac ... in_args = 0, returned = 0, others = 0
wall time = 711

old:

Testing T -> U ... in_args = 0, returned = 0, others = 200
wall time = 770
Testing T, U ... in_args = 0, returned = 0, others = 200
wall time = 419
Testing T -> T ... in_args = 0, returned = 0, others = 200
wall time = 433
Testing crc32 ... in_args = 0, returned = 0, others = 0
wall time = 632
Testing spi::pac ... in_args = 0, returned = 0, others = 0
wall time = 729
```

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:1688]: Do you think it's worth trying to optimize it to reduce the gap on big queries or as a follow-up PR?

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:1688]: Not yet! Sorry, until https://github.com/rust-lang/rust/pull/126245 is ready, I won't have much time to take a look. Hopefully I should be "freed" from in a few hours. Then I'll gladly give a try at optimizing it away. I'm sure we can reduce the gap a little bit. (Your work on this feature remains amazing, can't wait for it to be merged!)

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:1688]: Ok I'm now "free". I'll take a look tomorrow or monday. :D

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:1688]: > Have you managed to run the profiler suite? It's at https://gitlab.com/notriddle/rustdoc-js-profile

After modifying the `NEW` env variable to target my local build, I can run everything, except I get:

```
ReferenceError: atob is not defined
    at new RoaringBitmap (/home/imperio/rust/rustdoc-js-profile/tmp/arti/toolchain_new/doc/static.files/search-78d69af6d344b93d.js:5:13812)
    at buildIndex (/home/imperio/rust/rustdoc-js-profile/tmp/arti/toolchain_new/doc/static.files/search-78d69af6d344b93d.js:5:16653)
    at Object.initSearch (/home/imperio/rust/rustdoc-js-profile/tmp/arti/toolchain_new/doc/static.files/search-78d69af6d344b93d.js:5:22650)
    at loadSearchJS (/home/imperio/rust/rustdoc-js-profile/src/tester.js:225:38)
    at main (/home/imperio/rust/rustdoc-js-profile/src/tester.js:311:28)
    at Object.<anonymous> (/home/imperio/rust/rustdoc-js-profile/src/tester.js:331:1)
    at Module._compile (internal/modules/cjs/loader.js:1085:14)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
    at Module.load (internal/modules/cjs/loader.js:950:32)
    at Function.Module._load (internal/modules/cjs/loader.js:790:12)
```

Apparently the function was removed from nodejs so needs to implement it ourselves as follows:

```js
function atob(str) {
    return Buffer.from(encodedText, 'base64').toString('utf8');
}
```

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:1688]: There doesn't seem to be any error but the flamegraph is almost empty and the other files are empty. I suppose it failed somewhere but didn't provide the information... I'll continue my profiling using firefox directly.

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:1688]: This is my diff btw:

```diff
diff --git a/src/main.rs b/src/main.rs
index 5f938b3..c09ac55 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -163,7 +163,7 @@ fn main() {
             .output()
             .expect("run cargo doc on new toolchain");
         if !out.status.success() {
-            panic!("stdout = {stdout:?}, stderr = {stderr:?}", stdout = String::from_utf8_lossy(&out.stdout), stderr = String::from_utf8_lossy(&out.stderr));
+            panic!("stdout = {stdout}\n\n\nstderr = {stderr}", stdout = String::from_utf8_lossy(&out.stdout), stderr = String::from_utf8_lossy(&out.stderr));
         }
         // Run old under CPU profiler 0x
         let mut cmd = Command::new("../../node_modules/.bin/0x");
diff --git a/src/tester.js b/src/tester.js
index bf8d550..7aa195d 100644
--- a/src/tester.js
+++ b/src/tester.js
@@ -4,7 +4,10 @@ const path = require("path");
 function loadContent(content) {
     const Module = module.constructor;
     const m = new Module();
-    m._compile(content, "tmp.js");
+    const extra = "function atob(str) {
+    return Buffer.from(encodedText, 'base64').toString('utf8');
+}\n";
+    m._compile(extra + content, "tmp.js");
     m.exports.ignore_order = content.indexOf("\n// ignore-order\n") !== -1 ||
         content.startsWith("// ignore-order\n");
     m.exports.exact_check = content.indexOf("\n// exact-check\n") !== -1 ||
```

**Comment by GuillaumeGomez** [comment]
Found a bug (`undefined` in query match explanations):

image

**Comment by GuillaumeGomez** [comment]
Sorry, I misread it. So no, it's great as is. :+1:

File: src/librustdoc/html/static/js/search.js [hunk 20]
```diff
@@ -15,6 +15,25 @@ if (!Array.prototype.toSpliced) {
...
 }
 
+function takeFromArray(array, count) {
+    const result = [];
+    while (count > 0) {
+        result.push(array.shift());
+        count -= 1;
+    }
+    return result;
+}
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:25]:
```suggestion
function takeFromArray(array, count) {
    return array.splice(0, count);
}
```

Is at least twice faster based on https://jsbench.me/.

File: src/librustdoc/html/static/js/search.js [hunk 21]
```diff
@@ -1382,16 +1428,338 @@ function initSearch(rawSearchIndex) {
...
+                                mgens = mgensOut;
+                                return true;
+                            },
+                            0,
+                        );
+                        return !!fnOutput;
+                    },
+                    0,
+                );
+            }
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:1497]:
```suggestion
            if (typeInfo !== "elems" && typeInfo !== "returned") {
                fnInputs = unifyFunctionTypes(
                    obj.type.inputs,
                    parsedQuery.elems,
                    obj.type.where_clause,
                    null,
                    mgensScratch => {
                        fnOutput = unifyFunctionTypes(
                            obj.type.output,
                            parsedQuery.returned,
                            obj.type.where_clause,
                            mgensScratch,
                            mgensOut => {
                                mgens = mgensOut;
                                return true;
                            },
                            0,
                        );
                        return !!fnOutput;
                    },
                    0,
                );
            } else {
                const arr = typeInfo === "elems" ? obj.type.inputs : obj.type.output;
                fnInputs = unifyFunctionTypes(
                    arr,
                    parsedQuery.elems,
                    obj.type.where_clause,
                    null,
                    mgensOut => {
                        mgens = mgensOut;
                        return true;
                    },
                    0,
                );
            }
```

is a bit shorter

**Comment by GuillaumeGomez** [comment]
So just one comment to improve performance on one function. Once applied, can you provide the new numbers please?

File: src/librustdoc/html/static/js/search.js (hunk 20 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js]: Sounds good to me as well!

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Looks good to me! Thanks for going along. :)

r=me once CI pass (or does it need an FCP? Not sure...)

**Comment by GuillaumeGomez** [comment]
Wouldn't `rustdoc-ui` be enough?

**Comment by GuillaumeGomez** [comment]
I was hoping it would. But you're right, I don't think it was set up yet.

**Comment by GuillaumeGomez** [comment]
It does! \o/

**Comment by camelid** [comment]
@rfcbot reviewed
@rfcbot concern bug with type parameters

Awesome work! The design for this is great, but it seems the type parameter information is incorrect:

image

`T` is `Z`. `A` is the allocator type and is not used in the type signature.

**Comment by camelid** [comment]
It also seems a little unfortunate that rustdoc search allows using different names for type parameters that are actually the same, like how I used both `T` and `Z` in the example above. Because some functions actually *do* let you turn one type into a completely different type (e.g., `Iterator::map`), but this also shows functions that only do identity transformations. But this existed prior to this PR, so it shouldn't block it of course.

**Comment by camelid** [comment]
An example of an even more confusing result:

image

**Comment by jsha** [comment]
This is very nice work, and I think it's a big improvement over the status quo, so I'm checking off my FCP box. But I do have some UI feedback if you'd like!

In general we should avoid adding new hover text / popover interactions. This one in particular is confusing because the icon used is the same as "Notable traits." Also, more minorly, it's inconsistent to put the `(i)` on the left side when we put it on the right for notable traits.

I'd propose to put the contents of "About this result" directly in the search result row. It will mean each row is rather tall, but we're already making them taller by adding the type signature. We should concede that type-driven searches are different from name-based searches and will have a UI that looks significantly different. For instance we can get rid of `method` at the front because it's a given.

Also, I find the "A is U" or "Iterator::Item is T" language confusing, in part because of the symmetry and in part because the order seems wrong. How about: "U matches **A**", "T matches **Iterator::Item**". That way the thing the user typed on the query is on the left, and the thing it matched against is on the right, in a natural query->result or cause->effect pattern. And bolding the second term resonates with how matched terms are bolded elsewhere in the type signature.

One last suggestion: What if we removed the feature of requiring `T` and `Z` to match distinct types? I get that it's more expressive, but it's unusual, and it interacts poorly with other features like reordering and defaulted type parameters (like Vec's Allocator parameter, which people don't often think about).

By comparison, in regex if I wrote `(.+) => (.+)`, both of those placeholders could match an identical string. In algebra, if I wrote `x + y = 2`, it would be valid for x and y to be the same. In Rust, if I wrote `struct Point<T, U>`, it would be valid for T and U to be the same type. Indeed that example is [used as an example in The Rust Programming Language section on generics](https://doc.rust-lang.org/book/ch10-01-syntax.html#in-struct-definitions).

**Comment by jsha** [comment]
> Are you also on board with getting rid of generics reordering, as described in [2024-07-08's meeting](https://rust-lang.zulipchat.com/#narrow/stream/393423-t-rustdoc.2Fmeetings/topic/meeting.202024-07-08/near/449965149)? I can open a single PR with both changes.

Yep!

**Comment by GuillaumeGomez** [comment]
cc @camelid

**Comment by alper** [comment]
Saw this and 1. I find Rustdoc absurdly off-putting to use 2. I had no idea that you could even do searches like this in that box.
